editor/components/font-sizes/with-font-sizes.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * External dependencies
+ */
+import { find, get, reduce, upperFirst } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { createHigherOrderComponent, Component, compose } from '@wordpress/element';
+import { withSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { getFontSize, getFontSizeClass } from './utils';
+
+const DEFAULT_FONT_SIZES = [];
+
+/**
+ * Higher-order component, which handles font size logic for class generation,
+ * font size value retrieval, and font size change handling.
+ *
+ * @param {...(object|string)} args The arguments should all be strings
+ *                                  Each string contains the font size attribute name e.g: 'fontSize'.
+ *
+ * @return {Function} Higher-order component.
+ */
+export default ( ...fontSizeNames ) => {
+	return createHigherOrderComponent(
+		compose( [
+			withSelect( ( select ) => {
+				const settings = select( 'core/editor' ).getEditorSettings();
+				return {
+					fontSizes: get( settings, [ 'fontSizes' ], DEFAULT_FONT_SIZES ),
+				};
+			} ),
+			( WrappedComponent ) => {
+				return class extends Component {
+					constructor( props ) {
+						super( props );
+
+						this.setters = this.createSetters();
+
+						this.state = {};
+					}
+
+					createSetters() {
+						return reduce( fontSizeNames, ( settersAccumulator, fontSizeAttributeName ) => {
+							const upperFirstFontSizeAttributeName = upperFirst( fontSizeAttributeName );
+							const customFontSizeAttributeName = `custom${ upperFirstFontSizeAttributeName }`;
+							settersAccumulator[ `set${ upperFirstFontSizeAttributeName }` ] =
+								this.createSetFontSize( fontSizeAttributeName, customFontSizeAttributeName );
+							return settersAccumulator;
+						}, {} );
+					}
+
+					createSetFontSize( fontSizeAttributeName, customFontSizeAttributeName ) {
+						return ( fontSizeValue ) => {
+							const fontSizeObject = find( this.props.fontSizes, { size: fontSizeValue } );
+							this.props.setAttributes( {
+								[ fontSizeAttributeName ]: fontSizeObject && fontSizeObject.slug ? fontSizeObject.slug : undefined,
+								[ customFontSizeAttributeName ]: fontSizeObject && fontSizeObject.slug ? undefined : fontSizeValue,
+							} );
+						};
+					}
+
+					static getDerivedStateFromProps( { attributes, fontSizes }, previousState ) {
+						return reduce( fontSizeNames, ( newState, fontSizeAttributeName ) => {
```

Comment:
This method will be called every time any prop changes which means every time any attribute will change. Is there any way to opimize it to avoid doing all those computations where values change for attributes that have nothing to do with fonts?

Reply:
Hi @gziolo, I optimized the performance of this function. And now in the most common case where props unrelated to font size change we just compare the font size props with the previous version. Of course, the code is now a little bit more complex because of this optimization.